File: packages/engine/src/lib/game-engine.ts
// engine/GameEngine.ts
// Core game engine - pure reducer that transforms state via actions

import { actionValidator } from './action-validator';
import { handleSwapCard } from './cases/swap-card';
import { handleCallVinto } from './cases/call-vinto';
import { handleConfirmPeek } from './cases/confirm-peek';
import { handleDeclareKingAction } from './cases/declare-king-action';
import { handleDiscardCard } from './cases/discard-card';
import { handleDrawCard } from './cases/draw-card';
import { handleExecuteQueenSwap } from './cases/execute-queen-swap';
import { handleFinishSetup } from './cases/finish-setup';
import { handleFinishTossInPeriod } from './cases/finish-toss-in';
import { handleParticipateInTossIn } from './cases/participate-in-toss';
import { handlePlayerTossInFinished } from './cases/player-toss-in-finished';
import { handlePeekSetupCard } from './cases/peek-setup-card';
import { handleProcessAITurn } from './cases/process-ai-turn';
import { handleSelectActionTarget } from './cases/select-action-target';
import { handleSetCoalitionLeader } from './cases/set-coalition-leader';
import { handleSkipQueenSwap } from './cases/skip-queen-swap';
import { handlePlayDiscard } from './cases/play-discard';
import { handleUseCardAction } from './cases/use-card';
import { handleUpdateDifficulty } from './cases/update-difficulty';
import { handleUpdateBotVersion } from './cases/update-bot-version';
import { handleSetNextDrawCard } from './cases/set-next-draw-card';
import { handleSwapHandWithDeck } from './cases/swap-hand-with-deck';
import { GameAction, GameState, NeverError } from '@vinto/shapes';
import { handleExecuteJackSwap } from './cases/execute-jack-swap';
import { handleSkipJackSwap } from './cases/skip-jack-swap';
import { advanceTurnAfterTossIn } from './utils/toss-in-utils';
import { handleSkipPeek } from './cases/skip-peek';
import { handleEmpty } from './cases/empty';

type GameEngineReduceResult =
  | {
      success: true;
      state: GameState;
    }
  | {
      success: false;
      state: GameState;
      reason: string;
    };
/**
 * GameEngine - The authoritative game logic
 *
 * This is a pure, stateless service that implements the core game rules.
 * It has ONE primary method: reduce()
 *
 * Key principles:
 * - Pure functions only (no side effects)
 * - No mutations (uses deep copy)
 * - No dependencies on UI, stores, or client code
 * - Fully testable
 * - Server-runnable
 */
export class GameEngine {
  /**
   * Core reducer: State + Action → New State
   *
   * This is the ONLY method that modifies game state
   * All game logic flows through here
   *
   * @param state Current game state
   * @param action Action to apply
   * @returns New game state (never mutates input)
   */
  static reduce(state: GameState, action: GameAction): GameEngineReduceResult {
    // Validate action is legal in current state
    const validation = actionValidator(state, action);
    if (!validation.valid) {
      return { success: false, state, reason: validation.reason }; // Return unchanged state for invalid actions
    }

    console.log(`[GameEngine] Processing action: ${action.type}`);
    // Route to specific handler based on action type
    let newState: GameState;
    switch (action.type) {
      case 'DRAW_CARD':
        newState = handleDrawCard(state, action);
        break;

      case 'SWAP_CARD':
        newState = handleSwapCard(state, action);
        break;

      case 'DISCARD_CARD':
        newState = handleDiscardCard(state, action);
        break;

      case 'PLAY_DISCARD':
        newState = handlePlayDiscard(state, action);
        break;

      case 'USE_CARD_ACTION':
        newState = handleUseCardAction(state, action);
        break;

      case 'SELECT_ACTION_TARGET':
        newState = handleSelectActionTarget(state, action);
        break;

      case 'CONFIRM_PEEK':
        newState = handleConfirmPeek(state, action);
        break;

      case 'SKIP_PEEK':
        newState = handleSkipPeek(state, action);
        break;

      case 'CALL_VINTO':
        newState = handleCallVinto(state, action);
        break;

      case 'EXECUTE_JACK_SWAP':
        newState = handleExecuteJackSwap(state, action);
        break;

      case 'SKIP_JACK_SWAP':
        newState = handleSkipJackSwap(state, action);
        break;

      case 'EXECUTE_QUEEN_SWAP':
        newState = handleExecuteQueenSwap(state, action);
        break;

      case 'SKIP_QUEEN_SWAP':
        newState = handleSkipQueenSwap(state, action);
        break;

      case 'DECLARE_KING_ACTION':
        newState = handleDeclareKingAction(state, action);
        break;

      case 'PARTICIPATE_IN_TOSS_IN':
        newState = handleParticipateInTossIn(state, action);
        break;

      case 'PLAYER_TOSS_IN_FINISHED':
        newState = handlePlayerTossInFinished(state, action);
        break;

      case 'FINISH_TOSS_IN_PERIOD':
        newState = handleFinishTossInPeriod(state, action);
        break;

      case 'SET_COALITION_LEADER':
        newState = handleSetCoalitionLeader(state, action);
        break;

      case 'PROCESS_AI_TURN':
        newState = handleProcessAITurn(state, action);
        break;

      case 'PEEK_SETUP_CARD':
        newState = handlePeekSetupCard(state, action);
        break;

      case 'FINISH_SETUP':
        newState = handleFinishSetup(state, action);
        break;

      case 'UPDATE_DIFFICULTY':
        newState = handleUpdateDifficulty(state, action);
        break;

      case 'UPDATE_BOT_VERSION':
        newState = handleUpdateBotVersion(state, action);
        break;

      case 'SET_NEXT_DRAW_CARD':
        newState = handleSetNextDrawCard(state, action);
        break;

      case 'SWAP_HAND_WITH_DECK':
        newState = handleSwapHandWithDeck(state, action);
        break;

      case 'EMPTY':
        newState = handleEmpty(state, action);
        break;

      default:
        throw new NeverError(action);
    }

    if (newState === state) {
      return {
        success: false,
        state,
        reason: `Action handler for ${action.type} did not modify state`,
      };
    }

    // POST-ACTION PROCESSING
    // Auto-advance turn if toss-in queue is empty and all players are ready
    if (
      newState.activeTossIn &&
      newState.activeTossIn.queuedActions.length === 0 &&
      newState.pendingAction === null &&
      newState.activeTossIn.playersReadyForNextTurn.length ===
        newState.players.length &&
      action.type !== 'EMPTY' // skip for empty action
    ) {
      // All queued actions have been processed
      // Advance turn automatically
      advanceTurnAfterTossIn(newState, `GameEngine.reduce(${action.type})`);
    }

    return { success: true, state: newState };
  }
}

File: kmp/shared/engine/src/commonMain/kotlin/game/vinto/engine/GameEngine.kt
package game.vinto.engine

import game.vinto.engine.cases.handleConfirmPeek
import game.vinto.engine.cases.handleDiscardCard
import game.vinto.engine.cases.handleDrawCard
import game.vinto.engine.cases.handleEmpty
import game.vinto.engine.cases.handleEndRound
import game.vinto.engine.cases.handleFinishSetup
import game.vinto.engine.cases.handleFinishTossInPeriod
import game.vinto.engine.cases.handlePeekSetupCard
import game.vinto.engine.cases.handlePlayDiscard
import game.vinto.engine.cases.handleSwapCard
import game.vinto.engine.cases.handleUseCardAction
import game.vinto.engine.cases.handleProcessAiTurn
import game.vinto.engine.cases.handleExecuteJackSwap
import game.vinto.engine.cases.handleExecuteQueenSwap
import game.vinto.engine.cases.handleCallVinto
import game.vinto.engine.cases.handleDeclareKingAction
import game.vinto.engine.cases.handleParticipateInTossIn
import game.vinto.engine.cases.handlePlayerTossInFinished
import game.vinto.engine.cases.handleSelectActionTarget
import game.vinto.engine.cases.handleSkipJackSwap
import game.vinto.engine.cases.handleSkipQueenSwap
import game.vinto.engine.cases.handleSetCoalitionLeader
import game.vinto.engine.cases.handleSetNextDrawCard
import game.vinto.engine.cases.handleSkipPeek
import game.vinto.engine.cases.handleSwapHandWithDeck
import game.vinto.engine.cases.handleUpdateDifficulty
import game.vinto.shapes.Card
import game.vinto.shapes.GameAction
import game.vinto.shapes.GameState

/** The outcome of a reduction. An invalid action leaves the state untouched. */
/**
 * A card the table was shown, and by whom it is held.
 *
 * The rules turn a card face up in two places — a King's declaration and a failed toss-in —
 * and in both the card stays in the hand it was in. It is public for that moment and private
 * again afterwards, which makes it an *event* rather than a fact about the game: this is what
 * a room would send its four clients, and what this engine hands back to the one it has.
 */
data class PublicReveal(val playerId: String, val position: Int, val card: Card)

sealed interface ReduceResult {
    val state: GameState

    data class Success(
        override val state: GameState,
        /** What this action turned face up for everybody. Usually nothing. */
        val revealed: List<PublicReveal> = emptyList(),
    ) : ReduceResult
    data class Failure(override val state: GameState, val reason: String) : ReduceResult
}

/**
 * The authoritative game logic: state + action -> new state.
 *
 * Pure and stateless. The mutation inside is confined to a working copy that never escapes
 * this call — see [MutableGameState].
 *
 * Ported from `packages/engine/src/lib/game-engine.ts`.
 */
object GameEngine {

    fun reduce(state: GameState, action: GameAction): ReduceResult {
        val validation = ActionValidator.validate(state, action)
        if (validation is Validation.Invalid) {
            return ReduceResult.Failure(state, validation.reason)
        }

        val working = state.toMutable()
        val changed = dispatch(working, action)

        if (!changed) {
            return ReduceResult.Failure(
                state,
                "Action handler for ${action.type} did not modify state",
            )
        }

        if (shouldAdvanceTurn(working, action)) {
            advanceTurnAfterTossIn(working)
        }

        return ReduceResult.Success(working.freeze(), working.revealed.toList())
    }

    /**
     * Routes an action to its handler.
     *
     * Detekt reads this as highly complex; what it is measuring is the size of the action
     * union, not the difficulty of the code. An exhaustive `when` over the sealed hierarchy
     * is what makes adding an action a compile error at this site — and it is why there is
     * no `else` branch: every action has a handler, and the compiler holds that true.
     */
    @Suppress("CyclomaticComplexMethod")
    private fun dispatch(working: MutableGameState, action: GameAction): Boolean =
        when (action) {
            is GameAction.DrawCard -> handleDrawCard(working, action)
            is GameAction.DiscardCard -> handleDiscardCard(working, action)
            is GameAction.ConfirmPeek -> handleConfirmPeek(working, action)
            is GameAction.SkipPeek -> handleSkipPeek(working, action)
            is GameAction.EndRound -> handleEndRound(working)

            is GameAction.SetCoalitionLeader -> handleSetCoalitionLeader(working, action)
            is GameAction.ProcessAiTurn -> handleProcessAiTurn(working, action)
            is GameAction.PeekSetupCard -> handlePeekSetupCard(working, action)
            is GameAction.FinishSetup -> handleFinishSetup(working, action)
            is GameAction.UpdateDifficulty -> handleUpdateDifficulty(working, action)
            is GameAction.Empty -> handleEmpty(working, action)
            is GameAction.PlayDiscard -> handlePlayDiscard(working, action)
            is GameAction.UseCardAction -> handleUseCardAction(working, action)
            is GameAction.SwapCard -> handleSwapCard(working, action)
            is GameAction.SelectActionTarget -> handleSelectActionTarget(working, action)
            is GameAction.PlayerTossInFinished -> handlePlayerTossInFinished(working, action)
            is GameAction.ParticipateInTossIn -> handleParticipateInTossIn(working, action)
            is GameAction.DeclareKingAction -> handleDeclareKingAction(working, action)
            is GameAction.CallVinto -> handleCallVinto(working, action)
            is GameAction.FinishTossInPeriod -> handleFinishTossInPeriod(working, action)
            is GameAction.SetNextDrawCard -> handleSetNextDrawCard(working, action)
            is GameAction.SwapHandWithDeck -> handleSwapHandWithDeck(working, action)
            is GameAction.ExecuteJackSwap -> handleExecuteJackSwap(working, action)
            is GameAction.SkipJackSwap -> handleSkipJackSwap(working, action)
            is GameAction.ExecuteQueenSwap -> handleExecuteQueenSwap(working, action)
            is GameAction.SkipQueenSwap -> handleSkipQueenSwap(working, action)
        }

    /**
     * Whether the turn should advance now: the toss-in queue has drained, nothing is
     * pending, and every player has confirmed. EMPTY is excluded so a no-op cannot end a
     * turn.
     */
    private fun shouldAdvanceTurn(working: MutableGameState, action: GameAction): Boolean {
        if (action is GameAction.Empty) return false
        val tossIn = working.activeTossIn ?: return false

        return tossIn.queuedActions.isEmpty() &&
            working.pendingAction == null &&
            tossIn.playersReadyForNextTurn.size == working.players.size
    }
}
